Add release 2.6.1 blog for website (#7756)
apache/pulsar: site2/website/blog/2020-08-17-Apache-Pulsar-2-6-1.md

---
author: XiaoLong Ran
authorURL: https://twitter.com/wolf4j1
title: Apache Pulsar 2.6.1
---
We are very glad to see the Apache Pulsar community has successfully released the wonderful 2.6.1 version after accumulated hard work. It is a great milestone for this fast-growing project and the whole Pulsar community. This is the result of a huge effort from the community, with over 90 commits and a long list of improvements and bug fixes.

Here is a selection of some of the most interesting and major features added to Pulsar 2.6.1.

<!--truncate-->

## Broker

- **Limit the batch size to the minimum of the `maxNumberOfMessages` and `maxSizeOfMessages`**

The Batch size is not limited to the minimum of the `maxNumberOfMessages` and `maxSizeOfMessages` from the `BatchReceive` policy.

1. Batch size is not limited to the minimum of the maxNumberOfMessages and maxSizeOfMessages from the BatchReceive policy.

> 2. When the batch size is greater than the receiveQ of the consumer (I used a batch size of 3000 and a receiveQ of 500) I notice the following issues:

Reviewer: ```suggestion
2. When the batch size is greater than the receiveQ of the consumer (I used a batch size of 3000 and a receiveQ of 500), I notice the following issues:
```

	a. In a multiTopic (pattern) consumer the client stops receiving any messages I think it getting paused and never resumed when setting a timeout in the batch policy, only one batch is fetched and the client never resumed.

For more information about implementation details, see [PR-6865](https://github.com/apache/pulsar/pull/6865).

- **Fix hash range conflict issue in Key_Shared with sticky hash range**

> In `Key_Shared` subscription using `stickyHashRange`, consumers interleaving hashes are not allowed. For example, consumer 1 hash: [[0, 99], [400, 65535]], consumer 2 hash: [[100,399]].

Reviewer: ```suggestion
In `Key_Shared` subscription where the`stickyHashRange` is used, consumers are not allowed to use interleaving hashes.
```

The pull request will fix hash range conflict issue in `Key_Shared` with sticky hash range.

For more information about implementation details, see [PR-7231](https://github.com/apache/pulsar/pull/7231).

- **Fix: get lookup permission error**

> Currently，when pulsar AuthorizationService check lookup permission, if the user has the role canProducer **or** canConsumer means it can perform canLookup operations, but actually in the code:

Reviewer: ```suggestion
Currently，when Pulsar AuthorizationService checks lookup permission, if the user has the role canProducer **or** canConsumer, it means that the user can perform canLookup operations, but actually in the code:
```

> Currently，when pulsar AuthorizationService check lookup permission, if the role canProducer **or** canConsumer mean that canLookup, but actually in the code:

Reviewer: this sentence have some grammatical error. please re-write this sentence.

```java
try {
    return canLookupAsync(topicName, role, authenticationData)
            .get(conf.getZooKeeperOperationTimeoutSeconds(), SECONDS);
}
```

> if the method `canProduce` or `canConsume` throw exception, `canLookup` will just throw the exception and won't check the other permission.

Reviewer: ```suggestion
If the method `canProduce` or `canConsume` throw an exception,  the `canLookup` method will just throw the exception and will not check the other permission.
```

> If the method `canProduce` or `canConsume` throw an exception,  the `canLookup` method will just throw the exception and will not check the other permission.

Reviewer: ```suggestion
If the method `canProduce` or `canConsume` throw an exception, the `canLookup` method will just throw the exception and will not check the other permission.
```

```

For more information about implementation details, see [PR-7495](https://github.com/apache/pulsar/pull/7495).

- **Fix update partitions error for non-persistent topic**

When updating partitions on a non-persistent topic, Error 409 is returned. The pull request will fix partitions error for non-persistent topic.

For more information about implementation details, see [PR-7459](https://github.com/apache/pulsar/pull/7459).

## Zookeeper

- **Use hostname for bookie rack awareness mapping**

In [PR-5607](https://github.com/apache/pulsar/pull/5607), the `useHostName()` was added with `return false`. That means that the rack-aware policy will try to resolve the Bookie's hostname into an IP address and then use that IP address to figure out to which rack the bookie belongs.

There are 2 problems: 
 1. The IP won't match the hostname which is recorded in the `/bookies` z-node
 2. If there is an error in resolving the bookie hostname (eg: transient DNS error), an NPE exception will be triggered and the BK client will never realize that this bookie was ever seen as available in the cluster.

> The exception is thrown at 77, since `getAddress()` yealds a `null` given that the address is unresolved. 

Reviewer: ```suggestion
The exception is thrown at Line 77, since `getAddress()` returns a `null` given that the address is unresolved. 
```

```java
74        if (dnsResolver.useHostName()) {
75            names.add(addr.getHostName());
76        } else {
77            names.add(addr.getAddress().getHostAddress());
78        }
```

The default implementation for the `DnsResolver.useHostName()` is to return true.

For more information about implementation details, see [PR-7361](https://github.com/apache/pulsar/pull/7361).

## Java Client

- **Fix issue where HTTP header used in Athenz authentication can not be renamed**

The authentication plugin for Athenz allows users to change the name of the HTTP header for sending an authentication token to a broker server with a parameter named `roleHeader`. The change will hold the value of the `roleHeader` parameter on the `AuthenticationAthenz` side, and use it directly as the header name.
                                                                                                                                                                                                    
For more information about implementation details, see [PR-7311](https://github.com/apache/pulsar/pull/7311).

> - **Fix batch ack set recycled multiple times**

Author: ack set

> - **Fix batch ack set recycled multiple times**

Reviewer: The following sentences use ackset instead of ack set. you should use consistent words all through the docs.

For more information about implementation details, see [PR-7615](https://github.com/apache/pulsar/pull/7615).

- **Fix batch index filter issue in Consumer**

Fix batch index filter issue in Consumer.

For more information about implementation details, see [PR-7654](https://github.com/apache/pulsar/pull/7654).

## CPP Client

- **CPP Oauth2 authentication client**

Pulsar supports authenticating clients using OAuth 2.0 access tokens. You can use tokens to identify a Pulsar client and associate with some "principal" (or "role") that is permitted to do some actions (eg: publish messages to a topic or consume messages from a topic). This change tries to support it in cpp client.

For more information about implementation details, see [PR-7467](https://github.com/apache/pulsar/pull/7467).

- **Fix partition index error in close callback**

In partitioned producer/consumer's close callback, the partition index is always 0. We need to pass `ProducerImpl/ConsumerImpl`'s internal partition index field to `PartitionedProducerImpl/PartitionedConsumerImpl`'s close callback.

For more information about implementation details, see [PR-7282](https://github.com/apache/pulsar/pull/7282).

> - **Fix segment crashes that caused by race condition of timer in cpp client**

Reviewer: ```suggestion
- **Fix segment crashes that caused by race condition of timer in CPP client**
```

---
author: XiaoLong Ran
authorURL: https://twitter.com/wolf4j1
title: Apache Pulsar 2.6.1
---
We are very glad to see the Apache Pulsar community has successfully released the wonderful 2.6.1 version after accumulated hard work. It is a great milestone for this fast-growing project and the whole Pulsar community. This is the result of a huge effort from the community, with over 90 commits and a long list of improvements and bug fixes.

Here is a selection of some of the most interesting and major features added to Pulsar 2.6.1.

<!--truncate-->

## Broker

#### Limit the batch size to the minimum of the `maxNumberOfMessages` and `maxSizeOfMessages`

The Batch size is not limited to the minimum of the `maxNumberOfMessages` and `maxSizeOfMessages` from the `BatchReceive` policy.

1. Batch size is not limited to the minimum of the maxNumberOfMessages and maxSizeOfMessages from the BatchReceive policy.

> 2. When the batch size is greater than the receiveQ of the consumer (I used a batch size of 3000 and a receiveQ of 500), I notice the following issues:

Reviewer 2: Take care of the following:
1. Item 1 is identical to the sentence above. 
2. It's release blog, could we avoid using "I used ..., I notice....)?

> 	In a multi-topic (pattern) consumer, the client stops receiving any messages. I think it gets paused and never resumed when setting a timeout in the batch policy. Only one batch is fetched and the client is never resumed.

Reviewer 2: what does "it" mean in "I think it gets paused..."?

For more information about implementation details, see [PR-6865](https://github.com/apache/pulsar/pull/6865).

#### Fix hash range conflict issue in Key_Shared subscription with sticky hash range

> In `Key_Shared` subscription where the`stickyHashRange` is used, consumers are not allowed to use interleaving hashes.

Reviewer 2: the`stickyHashRange` --> the `stickyHashRange`

Author: What is difference?

Reviewer 2: add a space between the two words.

> The pull request will fix hash range conflict issue in `Key_Shared` with sticky hash range.

Reviewer 2: will fix --> fixes

For more information about implementation details, see [PR-7231](https://github.com/apache/pulsar/pull/7231).

#### Fix: get lookup permission error

> Currently，when Pulsar AuthorizationService checks lookup permission, if the user has the role canProducer **or** canConsumer, it means that the user can perform canLookup operations, but actually in the code:

Reviewer 2: I. why do we use bold font for "**or**"?
2. Adopt code font or bold font for  the following role/operation/method.
`canProducer` or `canConsumer`
`canLookup`

```java
try {
    return canLookupAsync(topicName, role, authenticationData)
            .get(conf.getZooKeeperOperationTimeoutSeconds(), SECONDS);
}
```

> If the method `canProduce` or `canConsume` throw an exception, the `canLookup` method will just throw the exception and will not check the other permission.

Reviewer 2: Release notes and blogs describe some objective facts, use present tense, and avoid using "will" if possible.

return new PersistentSubscription(this, subscriptionName, cursor, false);
```

Additionally, the NPE leads to the topic usage count increasing to 1. When deleting a topic, the topic cannot be deleted even if you use the force flag.

For more information about implementation, see [PR-7355](https://github.com/apache/pulsar/pull/7355).

> ### Avoid  an NPE occurs in the `ManagedLedgerImpl.isOffloadedNeedsDelete` method

Reviewer 2: ```suggestion
### Avoid an NPE occurs in the `ManagedLedgerImpl.isOffloadedNeedsDelete` method
```

> ### Avoid  an NPE occurs in the `ManagedLedgerImpl.isOffloadedNeedsDelete` method

Reviewer 2: ```suggestion
### Avoid an NPE occurs in the `ManagedLedgerImpl.isOffloadedNeedsDelete` method
```

```
16:34:25.779 [pulsar-io-54-7:org.apache.bookkeeper.client.LedgerHandle@748] ERROR org.apache.bookkeeper.client.LedgerHandle - IncorrectParameterException on ledgerId:0 firstEntry:-1 lastEntry:-1
16:34:25.779 [pulsar-client-io-82-1:org.apache.pulsar.client.impl.ConsumerImpl@1986] INFO  org.apache.pulsar.client.impl.ConsumerImpl - [persistent://external-repl-prop/pulsar-function-admin/assignment][c-use-fw-localhost-0-function-assignment-initialize-reader-b21f7607c9] Successfully getLastMessageId 0:-1
16:34:25.779 [pulsar-client-io-82-1:org.apache.pulsar.client.impl.ClientCnx@602] WARN  org.apache.pulsar.client.impl.ClientCnx - [id: 0xc78f4a0e, L:/127.0.0.1:55657 - R:localhost/127.0.0.1:55615] Received error from server: Failed to get batch size for entry org.apache.bookkeeper.mledger.ManagedLedgerException: Incorrect parameter input
16:34:25.779 [pulsar-client-io-82-1:org.apache.pulsar.client.impl.ClientCnx@612] WARN  org.apache.pulsar.client.impl.ClientCnx - [id: 0xc78f4a0e, L:/127.0.0.1:55657 - R:localhost/127.0.0.1:55615] Received unknown request id from server: 10
```

> PR-7495 adds a return statement to code, so GetLastEntry()  reads the last entry, instead of `-1`.  

Reviewer 2: ```suggestion
PR-7495 adds a return statement to code, so GetLastEntry() reads the last entry, instead of `-1`.  
```

> PR-7495 adds a return statement to code, so GetLastEntry()  reads the last entry, instead of `-1`.  

Reviewer 2: ```suggestion
PR-7495 adds a return statement to code, so GetLastEntry() reads the last entry, instead of `-1`.  
```

78        }
```

The default implementation for the `DnsResolver.useHostName()` returns `true`.

For more information about implementation, see [PR-7361](https://github.com/apache/pulsar/pull/7361).

## Java Client

### Fix the issue that the HTTP header used in Athenz authentication can not be renamed

The authentication plugin for Athenz allows users to change the name of the HTTP header for sending an authentication token to a broker server with a parameter named `roleHeader`. The change uses the value of the `roleHeader` parameter on the `AuthenticationAthenz` side, and uses it directly as the header name.
                                                                                                                                                                                                    
For more information about implementation, see [PR-7311](https://github.com/apache/pulsar/pull/7311).

### Fix the issue that batch ack set is recycled multiple times

> The batch ack sets are recycled multiple times, due to race condition in group ack flush and cumulative Ack. So we add a recycled state check for the ack set in PR-7409,  and fix the recycle issue.

Reviewer 2: ```suggestion
The batch ack sets are recycled multiple times, due to race condition in group ack flush and cumulative Ack. So we add a recycled state check for the ack set in PR-7409, and fix the recycle issue.
```

> The batch ack sets are recycled multiple times, due to race condition in group ack flush and cumulative Ack. So we add a recycled state check for the ack set in PR-7409,  and fix the recycle issue.

Reviewer 2: ```suggestion
The batch ack sets are recycled multiple times, due to race condition in group ack flush and cumulative Ack. So we add a recycled state check for the ack set in PR-7409, and fix the recycle issue.
```

For more information about implementation, see [PR-7589](https://github.com/apache/pulsar/pull/7589).

### OAuth2 authentication plugin uses AsyncHttpClient

Previously, the OAuth2 client authentication plugin used Apache HTTP client lib to make requests, Apache HTTP client is used to validate hostname. As suggested in [#7612](https://github.com/apache/pulsar/issues/7612), we get rid of the dependency of using Apache HTTP client.

In PR-7615, OAuth2 client authentication plugin uses AsyncHttpClient, which is used in client and broker. For more information about implementation, see [PR-7615](https://github.com/apache/pulsar/pull/7615).

## CPP Client

### CPP Oauth2 authentication client

Pulsar supports authenticating clients using OAuth 2.0 access tokens. You can use tokens to identify a Pulsar client and associate with some "principal" (or "role") that is permitted to do some actions (eg: publish messages to a topic or consume messages from a topic). This change tries to support it in cpp client.

For more information about implementation, see [PR-7467](https://github.com/apache/pulsar/pull/7467).

### Fix partition index error in close callback

> In partitioned producer/consumer's close callback, the partition index is always `0`. The `ProducerImpl/ConsumerImpl`'s internal partition index field should be passed to `PartitionedProducerImpl/PartitionedConsumerImpl`'s close callback.

Reviewer 2: ```suggestion
In partitioned producer/consumer's close callback, the partition index is always `0`. The `ProducerImpl/ConsumerImpl` internal partition index field should be passed to `PartitionedProducerImpl/PartitionedConsumerImpl` close callback.
```

> In partitioned producer/consumer's close callback, the partition index is always `0`. The `ProducerImpl/ConsumerImpl`'s internal partition index field should be passed to `PartitionedProducerImpl/PartitionedConsumerImpl`'s close callback.

Reviewer 2: ```suggestion
In partitioned producer/consumer's close callback, the partition index is always `0`. The `ProducerImpl/ConsumerImpl` internal partition index field should be passed to `PartitionedProducerImpl/PartitionedConsumerImpl` close callback.
```

This is because the broker 2.5.0 supports "bookkeeperMetadataServiceUri" and the admin client returns a `null` field, thus causing the NPE.

For more information about implementation, see [PR-7528](https://github.com/apache/pulsar/pull/7528).

## pulsar-perf

### Support `tlsAllowInsecureConnection` in pulsar-perf produce/consume/read performance tests

Add `tlsAllowInsecureConnection` config to the CLI tool **pulsar-perf**, to support produce/consume/read performance tests to clusters with insecure TLS connections.

For more information about implementation, see [PR-7300](https://github.com/apache/pulsar/pull/7300).

## More information

- To download Apache Pulsar 2.6.1, click [download](https://pulsar.apache.org/en/download/).

> - For more information about Apache Pulsar 2.6.1, see [2.6.1 release notes](https://pulsar.apache.org/release-notes/#2.6.1 and [2.6.1 PR list](https://github.com/apache/pulsar/pulls?q=is%3Apr+label%3Arelease%2F2.6.1+is%3Aclosed).

Reviewer 2: ```suggestion
- For more information about Apache Pulsar 2.6.1, see [2.6.1 release notes](https://pulsar.apache.org/release-notes/#2.6.1) and [2.6.1 PR list](https://github.com/apache/pulsar/pulls?q=is%3Apr+label%3Arelease%2F2.6.1+is%3Aclosed).
```

> - For more information about Apache Pulsar 2.6.1, see [2.6.1 release notes](https://pulsar.apache.org/release-notes/#2.6.1 and [2.6.1 PR list](https://github.com/apache/pulsar/pulls?q=is%3Apr+label%3Arelease%2F2.6.1+is%3Aclosed).

Reviewer 2: ```suggestion
- For more information about Apache Pulsar 2.6.1, see [2.6.1 release notes](https://pulsar.apache.org/release-notes/#2.6.1) and [2.6.1 PR list](https://github.com/apache/pulsar/pulls?q=is%3Apr+label%3Arelease%2F2.6.1+is%3Aclosed).
```
